### PR TITLE
Dendrite Text Generate

### PR DESCRIPTION
**Pull request by unconst**
Adds text generation as a function on the dendrite:

# The below code prompts the best miner on the network.

import bittensor
wallet = bittensor.wallet()
graph = bittensor.metagraph().sync()
print ( bittensor.dendrite( wallet = wallet ).generate( 
        endpoints = graph.endpoints[graph.incentive.sort()[1][-1]], 
        prompt = "The quick brown fox jumped over the lazy dog", 
        num_to_generate = 20
    )
)

### CODE REVIEW

File: bittensor/_dendrite/dendrite_impl.py [hunk 1]
```diff
@@ -351,6 +351,114 @@ def _forward(
...
+            max_time (:obj: float, :default: 150): 
+                The maximum time that a server can use to generate
+            forward_request_serializer_type (:obj:`bittensor.proto.Serializer.Type` of shape :obj:`(1)`, `optional`, :default: `bittensor.proto.Serializer.MSGPACK`):
+                Serializer used to pack torch tensors on forward request.
+            forward_response_serializer_type (:obj:`bittensor.proto.Serializer.Type` of shape :obj:`(1)`, `optional`, :default: `bittensor.proto.Serializer.MSGPACK`):
+                Serializer used to pack torch tensors on forward response.
+            backward_request_serializer_type (:obj:`bittensor.proto.Serializer.Type` of shape :obj:`(1)`, `optional`, :default: `bittensor.proto.Serializer.MSGPACK`):
+                Serializer used to pack torch tensors on forward request.
+            backward_response_serializer_type (:obj:`bittensor.proto.Serializer.Type` of shape :obj:`(1)`, `optional`, :default: `bittensor.proto.Serializer.MSGPACK`):
+                Serialzer used to pack torch tensors on backward response.
```
Comment (camfairchild) [inline bittensor/_dendrite/dendrite_impl.py]: These aren't present in the function params

File: bittensor/_dendrite/dendrite_impl.py [hunk 2]
```diff
@@ -351,6 +351,114 @@ def _forward(
...
+                    temperature = temperature,
+                    repetition_penalty = repetition_penalty,
+                    length_penalty = length_penalty,
+                    max_time = max_time,
+                    num_beam_groups = num_beam_groups,
+                )
+            ],
+            timeout = timeout
+        )
+        generations = []
```
Comment (camfairchild) [inline bittensor/_dendrite/dendrite_impl.py:458]: `output, codes, times = response`
should we not verify no error codes before we attempt to decode the tokens.
Perhaps output a useful error message with the mapped error code message

**Review by camfairchild — CHANGES_REQUESTED** [review]
Minor changes. Otherwise LGTM

File: bittensor/_dendrite/dendrite_impl.py (hunk 1 above)
Comment (unconst) [reply bittensor/_dendrite/dendrite_impl.py]: Good catch

**Review by camfairchild — APPROVED** [review]
LGTM

File: bittensor/_dendrite/dendrite_impl.py [hunk 3]
```diff
@@ -351,6 +351,121 @@ def _forward(
...
+        # Parse responses to natural language.
+        generations = []
+        for text_tensor in response[0]:
+            generations.append( tokenizer.decode( text_tensor[0][0].long() ) )
+        codes = []
+        for code_tensor in response[1]:
+            codes.append( bittensor.utils.codes.code_to_string( code_tensor ) )
+        times = []
+        for time_tensor in response[2]:
+            times.append( time_tensor.item() )
```
Comment (eduardogr) [inline bittensor/_dendrite/dendrite_impl.py:466]: Each `for` is a nice candidate to be extracted into a new separated function in order to test them unitary:
- You get the correct index: 0, 1 or 2 depending on the function
- All elements were appended as expected
- you called the needed external function (that will be mocked): tokenizer.decode / bittensor.utils.codes.code_to_string / time_tensor.item

**Review by isabella618033 — APPROVED** [review]
LGTM

File: bittensor/_dendrite/dendrite_impl.py [hunk 4]
```diff
@@ -351,6 +351,121 @@ def _forward(
 
         return packed_outputs, packed_codes, packed_times
 
+
+    def generate(
```
Comment (eduardogr) [inline bittensor/_dendrite/dendrite_impl.py:355]: What's the intention of this function??

It's something to be used inside bittensor code / python code?. I'm wondering wether exposing an API would have sense

**Pull request merged**